pkg/controller/infrastructure/reconciler.go

```diff
@@ -94,42 +98,45 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 		return reconcile.Result{}, err
 	}
 
-	if infrastructure.DeletionTimestamp != nil {
-		return r.delete(r.ctx, infrastructure, cluster)
+	operationType := gardencorev1beta1helper.ComputeOperationType(infrastructure.ObjectMeta, infrastructure.Status.LastOperation)
+
+	switch {
+	case extensionscontroller.IsMigrated(infrastructure):
+		return reconcile.Result{}, nil
+	case operationType == gardencorev1beta1.LastOperationTypeMigrate:
+		return r.migrate(infrastructure, cluster)
+	case infrastructure.DeletionTimestamp != nil:
+		return r.delete(infrastructure, cluster)
+	case infrastructure.Annotations[gardencorev1beta1constants.GardenerOperation] == gardencorev1beta1constants.GardenerOperationRestore:
+		return r.restore(infrastructure, cluster, operationType)
+	default:
+		return r.reconcile(infrastructure, cluster, operationType)
 	}
-	return r.reconcile(r.ctx, infrastructure, cluster)
 }
 
-func (r *reconciler) reconcile(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure, cluster *extensionscontroller.Cluster) (reconcile.Result, error) {
-	if err := extensionscontroller.EnsureFinalizer(ctx, r.client, FinalizerName, infrastructure); err != nil {
+func (r *reconciler) reconcile(infrastructure *extensionsv1alpha1.Infrastructure, cluster *extensionscontroller.Cluster, operationType gardencorev1beta1.LastOperationType) (reconcile.Result, error) {
```

Comment:
Can you revert the ctx part here?

```suggestion
func (r *reconciler) reconcile(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure, cluster *extensionscontroller.Cluster, operationType gardencorev1beta1.LastOperationType) (reconcile.Result, error) {
```

```
Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx
```
Ref https://golang.org/pkg/context/

Reply:
thanks